### DISCUSSION TITLE
Render Targets and textures don’t get fully disposed on DirectX

### DISCUSSION THREAD

**Discussion by BitKovin** [MonoGame/MonoGame#8199] [Desktop computers]
DirectX implementation of Dispose method doesn’t really frees up memory. Here’s comparison of same program on directX and openGL
![image](https://github.com/MonoGame/MonoGame/assets/39824091/009da5b3-ab1d-43ec-b90a-62d2f091d9e1)
![image](https://github.com/MonoGame/MonoGame/assets/39824091/959b1ecf-6978-47a3-bb30-3cdf6ef57dfe)
While task manager and msi afterburner show decrease in memory usage program still kind of uses memory and will start take longer to create new render target and will eventually crash with “out of memory” error
![image](https://github.com/MonoGame/MonoGame/assets/39824091/fa35a3bb-9907-487d-a8b9-949500cc0db6)
This problem can be caused by simple disposing and creation of new render target after window gets resized.

**Comment by SimonDarksideJ** [comment]
So I've tested the above report on two clean DX and GL projects (code below) and have been unable to replicate the observation found.

Generating, Rendering and disposing of up to 100 RenderTargets per frame does appear to be cleaned up evenly between the two platforms, as shown below:

## DesktopGL (run on windows)
![image](https://github.com/MonoGame/MonoGame/assets/1793042/74498006-e157-41ea-b8e2-2c9c58369da3)

## Desktop DX
![image](https://github.com/MonoGame/MonoGame/assets/1793042/d50a2558-763e-4dc3-913d-32f03c3d2024)

The only notable difference between the two is that DX does seem to consume almost double the memory for the same amount of data between the two projects.

However, both platforms are disposing and clearing the data and not retaining it as far as I can tell.

---
### Sample code
```csharp
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using Microsoft.Xna.Framework.Input;

namespace perftest;

public class Game1 : Game
{
    private GraphicsDeviceManager _graphics;
    private SpriteBatch _spriteBatch;
    Texture2D grid;
    RenderTarget2D[] renderTargets;
    int maxRenderTargets = 100;
    Vector2 pos = new Vector2(200, 50);

    public Game1()
    {
        _graphics = new GraphicsDeviceManager(this);
        Content.RootDirectory = "Content";
        IsMouseVisible = true;
    }

    protected override void LoadContent()
    {
        _spriteBatch = new SpriteBatch(GraphicsDevice);

        grid = Content.Load<Texture2D>("grid");
        renderTargets = new RenderTarget2D[maxRenderTargets];
        CreateRenderTargets();
    }

    private void CreateRenderTargets()
    {
        for (int i = 0; i < maxRenderTargets; i++)
        {
            renderTargets[i] = new RenderTarget2D(
                               GraphicsDevice,
                               GraphicsDevice.PresentationParameters.BackBufferWidth,
                               GraphicsDevice.PresentationParameters.BackBufferHeight);
        }
    }

    protected override void Update(GameTime gameTime)
    {
        if (GamePad.GetState(PlayerIndex.One).Buttons.Back == ButtonState.Pressed || Keyboard.GetState().IsKeyDown(Keys.Escape))
            Exit();

        if(Keyboard.GetState().IsKeyDown(Keys.Space))
        {
            for (int i = 0; i < maxRenderTargets; i++)
            {
                ReleaseRenderTarget(renderTargets[i]);
            }
            CreateRenderTargets();
        }

        base.Update(gameTime);
    }

    protected override void Draw(GameTime gameTime)
    {
        for (int i = 0; i < maxRenderTargets; i++)
        {
            DrawRenderTarget(renderTargets[i]);
        }

        GraphicsDevice.Clear(Color.CornflowerBlue);
        _spriteBatch.Begin();
        for (int i = 0; i < maxRenderTargets; i++)
        {
            _spriteBatch.Draw(renderTargets[i], pos, Color.White);
        }
        _spriteBatch.End();
        base.Draw(gameTime);
    }

    private void DrawRenderTarget(RenderTarget2D renderTarget)
    {
        _graphics.GraphicsDevice.SetRenderTarget(renderTarget);

        _graphics.GraphicsDevice.Clear(Color.Black);

        _spriteBatch.Begin();
        _spriteBatch.Draw(grid, Vector2.Zero, Color.White);
        _spriteBatch.End();

        _graphics.GraphicsDevice.SetRenderTarget(null);
    }

    private void ReleaseRenderTarget(RenderTarget2D renderTarget)
    {
        renderTarget.Dispose();
    }
}
```

**Reply by BitKovin** [reply]
Damn. I'm very sorry. It was very dumb problem in my code. Problem was that I've initialized rendertarget and then replaces it's refenrece with another render target. But for some strange reason openGL disposed unreferenced rendertarget, but directX didn't. I'm sorry for Taking your time.

**Reply by SimonDarksideJ** [reply]
Don't worry about it, better if we test and validate our conclusions, to ensure there are no underlying issues that we ar unaware of.
